docs/RFCS/20210330_sql_stats_persistence.md

```diff
@@ -2,36 +2,741 @@
+CockroachDB node will check if there already exists the same fingerprint in the
+persisted system tables within the latest aggregation interval.
+* if such entry exists, the flush operation will aggregate the existing entry.
+* if such entry does not exist, the flush operation will insert a new entry.
+
+However, if we are filling up our memory buffer faster than we can flush them
+into system tables, and also since it is not desirable to block query
+execution, our only option here is to discard the new incoming statistics.
+We would be keeping track of number of statistics we have to discard using a
+counter and expose this as a metric. This allows administrators to monitor
+the health of the system.
+
+When DB Console issues fetch requests to CockroachDB node through HTTP endpoint,
+the code that handles the HTTP request will be updated to fetch the persisted
+statistics using follower read to minimize read-write contention. For the most
+up-to-date statistics, we would still need to utilize RPC fanout to retrieve the
+in-memory statistics from each node. However, the current implementation of our
+HTTP handler buffers all statistics it fetched from the cluster in memory
+before returning to the client. This can cause potential issue as we extend
+this HTTP endpoint to return the persisted statistics from the system table.
+The amount of persisted statistics can potentially exceed the available memory
+on the gateway node and cause the node to crash because of OOM. Therefore, we
+also need to update the existing HTTP endpoint to enable pagination. This way,
+we can prevent the HTTP handler from unboundedly buffering statistics in-memory
+when it reads from the system table.
+
+It is also worth noting that keeping the RPC fanout in this case will not make
+the existing situation worse since the response from RPC fanout will still only
+contain the statistics store in-memory in all other nodes.
+
+Additionally, we can implement a system view to allow users to access this
+up-to-date statistics within the SQL shell.
+
+Lastly, since the new system tables will be accessed frequently, in order to
+prevent [bottleneck](https://github.com/cockroachdb/cockroach/pull/63241) in
+name resolution, we want to cache the table descriptors for the new system
+tables.
+
+## Design Details
+
+### System table schema
+
+``` SQL
+CREATE TABLE system.sql_statement_stats (
+    aggregated_ts  TIMESTAMPTZ NOT NULL,
+    fingerprint_id BYTES NOT NULL,
+    app_name       STRING NOT NULL,
+    plan_hash      INT NOT NULL,
+    node_id        INT NOT NULL,
+
+    count        INT NOT NULL,
+    agg_interval INTERVAL NOT NULL,
+
+    metadata   JSONB NOT NULL,
+    /*
+    JSON Schema:
+    {
+      "$schema": "https://json-schema.org/draft/2020-12/schema",
+      "title": "system.sql_statement_stats.metadata",
+      "type": "object",
+      "properties": {
+        "sqlTyp":               { "type": "string" },
+        "query":                { "type": "string" },
```

Comment:
What is sqlTyp? And is this the same anonymized column in crdb_internal.node_statement_statistics?